kong/runloop/balancer.lua

```diff
@@ -164,15 +164,22 @@ local function apply_history(rb, history, start)
 
   for i = start, #history do
     local target = history[i]
+    local hostname = target.name
+    if history[i].upstream then
+      local upstream = get_upstream_by_id(history[i].upstream.id)
+      if upstream and upstream.hostname then
+        hostname = upstream.hostname
+      end
+    end
 
     if target.weight > 0 then
-      assert(rb:addHost(target.name, target.port, target.weight))
+      assert(rb:addHost(hostname, target.port, target.weight))
```

Comment:
One thing is confusing me a lot here. It was my understanding that the `upstream.hostname` attribute would only change the `Host` header sent with proxied requests, but not change the balancer logic otherwise. With this change, it seems that when `upstream.hostname` is set, this is no longer adding targets with their different IPs to the ring-balancer object (but instead is adding hostname-based entries to the object which are then subjected to DNS resolution).

I thought the idea was that, if I create a setup like this:

```
upstreams:
- name: myupstreaam
  hostname: example.com
  targets:
  - target: 10.0.0.1:80
  - target: 10.0.0.2:80
```

I want requests to be proxied either `10.0.0.1:80` or `10.0.0.2:80` with a header saying `Host: example.com` but not proxied to `example.com:80`.

Am I missing something?